metrics/metrics.go

```diff
@@ -106,11 +122,42 @@ func Start(metricsBind string, pprof http.Handler) {
 		fmt.Fprintf(w, "GC run complete")
 	})
 	s.Handler = mux
-	log.DefaultLogger.Warn("metrics", "listen finished", "err", s.ListenAndServe())
+	go func() {
+		log.DefaultLogger.Warn("metrics", "listen finished", "err", s.Serve(l))
+	}()
+	return l
 }
 
 // GroupHandler provides metrics shared to other group members
 // This HTTP handler, which would typically be mounted at `/metrics` exposes `GroupMetrics`
 func GroupHandler() http.Handler {
 	return promhttp.HandlerFor(GroupMetrics, promhttp.HandlerOpts{Registry: GroupMetrics})
 }
+
+// lazyPeerHandler is a structure that defers learning who current
+// group members are until an http request is received.
+type lazyPeerHandler struct {
+	peerHandler PeerHandler
+}
+
+func (l *lazyPeerHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	idx := strings.Replace(r.URL.Path, "/peer/", "", 1)
+	idxN, err := strconv.ParseUint(idx, 10, 64)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	handlers, err := l.peerHandler(r.Context())
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if idxN >= uint64(len(handlers)) {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	r.URL.Path = "/metrics"
```

Comment:
A small comments explaining why that's necessary is welcome ! I think i understand it's because you want to use the endpoint metrics/ but you could also create a fresh http request, why not doing that ?